{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Converting CSV to TSV\n",
    "The relatively raw CSV files manually generated from the XLSX files are not ideal for future steps. This script reads those in and generates uniform TSV files."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Import Packages & Setup Environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "import math"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def B721Allele(directory):\n",
    "    return(directory[:directory.find('_')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def inBlacklist(row, blacklist):\n",
    "    if row.loc['sequence'] in blacklist:\n",
    "        return(True)\n",
    "    return(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def removeBlacklist(df):\n",
    "    file = open('../../data/tsv/Blacklist.txt', 'r')\n",
    "    blacklist = set(file.read().split('\\n'))\n",
    "    return(df[df.apply(inBlacklist, axis=1, args=(blacklist, )) == False])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def numberPeptides(peptide):\n",
    "    return('Peptide:{:06}'.format(peptide.name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def updateSequence(sequence):\n",
    "    '''\n",
    "    Change modified amino acids to protein sequence\n",
    "    Expects a protein fasta string\n",
    "    Returns a list of corrected strings\n",
    "    '''\n",
    "    newSequence = ['']\n",
    "    for c in sequence:\n",
    "        if c == 'm':\n",
    "            for i in range(0, len(newSequence)):\n",
    "                newSequence[i] += 'M'\n",
    "        elif c == 'c':\n",
    "            for i in range(0, len(newSequence)):\n",
    "                newSequence[i] += 'M'\n",
    "        elif c == 'q':\n",
    "            newSequence = newSequence * 2\n",
    "            for i in range(0, len(newSequence)):\n",
    "                if i % 2 == 0:\n",
    "                    newSequence[i] += 'E'\n",
    "                else:\n",
    "                    newSequence[i] += 'Q'\n",
    "        else:\n",
    "            for i in range(0, len(newSequence)):\n",
    "                newSequence[i] += c\n",
    "    return(newSequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mapSequenceUpdate(row):\n",
    "    if isinstance(row.loc['sequenceMulti'], str):\n",
    "        seqMulti = row.loc['sequenceMulti'].split('|')\n",
    "        seqList = []\n",
    "        for seq in seqMulti:\n",
    "            seqList = seqList + updateSequence(seq)\n",
    "    else:\n",
    "        seq = row.loc['sequence']\n",
    "        seqList = updateSequence(seq)\n",
    "    seqList = list(set(seqList))\n",
    "    seqList = ','.join(seqList)\n",
    "    return(seqList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mhcGBMIFN(row):\n",
    "    directory = row.loc['directory']\n",
    "    IFN = directory.split('_')[-1]\n",
    "    if IFN == 'IFN':\n",
    "        return('+')\n",
    "    if IFN != 'IFN':\n",
    "        return('-')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def wholeGBMIFN(row):\n",
    "    TMT126 = row.loc['TMT_126_126.127']\n",
    "    TMT127 = row.loc['TMT_127_126.127']\n",
    "    if abs(TMT126 - TMT127) <= 0.2:\n",
    "        return('.')\n",
    "    if TMT126 > TMT127:\n",
    "        return('-')\n",
    "    if TMT127 > TMT126:\n",
    "        return('+')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## B721.221"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process B721 MHC-I on 15 alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "B721_15_MHC = pandas.read_csv(\n",
    "    '../../data/csv/B721.15.MHCI.csv',\n",
    "    sep=',',\n",
    "    header=0,\n",
    "    index_col=False,\n",
    "    low_memory=False)\n",
    "B721_15_MHC = removeBlacklist(B721_15_MHC)\n",
    "B721_15_MHC['allele'] = B721_15_MHC['directory'].apply(B721Allele)\n",
    "B721_15_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'parent_charge',\n",
    "        'subgroupSpecific',\n",
    "        'numLCrunsObserved',\n",
    "        'numKorR',\n",
    "        'variableSites',\n",
    "        'StartAA',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'missedCleavages',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'parent_m_over_z',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'totalIntensity',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'rank2Sequence',\n",
    "        'directory',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'sequenceMap',\n",
    "        'num_seqs_past_parent_filter'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process B721 MHC-I on 79 alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "B721_79_MHC = pandas.read_csv(\n",
    "    '../../data/csv/B721.79.MHCI.csv',\n",
    "    sep=',',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "B721_79_MHC = removeBlacklist(B721_79_MHC)\n",
    "B721_79_MHC['allele'] = B721_79_MHC['directory'].apply(B721Allele)\n",
    "B721_79_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'subgroupSpecific',\n",
    "        'numLCrunsObserved',\n",
    "        'missedCleavages',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'rank2Sequence',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'parent_m_over_z',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'sequenceMap',\n",
    "        'num_seqs_past_parent_filter'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge B721 MHC-I and assign peptide numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "B721_MHC = pandas.concat(\n",
    "    [B721_15_MHC, B721_79_MHC],\n",
    "    axis=0,\n",
    "    ignore_index=True,\n",
    "    sort=False)\n",
    "B721_MHC['Peptide:UID'] = B721_MHC.apply(numberPeptides, axis=1)\n",
    "B721_MHC['sequenceList'] = B721_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "B721_MHC.to_csv(\n",
    "    '../../data/tsv/B721.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process B721 Whole Proteome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "B721_Whole = pandas.read_csv(\n",
    "    '../../data/ssv/B721.Whole.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False,\n",
    "    low_memory=False\n",
    ")\n",
    "B721_Whole.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'subgroupSpecific',\n",
    "        'numLCrunsObserved',\n",
    "        'missedCleavages',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'rank2Sequence',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'parent_m_over_z',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'sequenceMap',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'nterm',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'next_aa',\n",
    "        'accession_number',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'recallMstag',\n",
    "        'numCcuts',\n",
    "        'numNCcuts',\n",
    "        'numIcuts',\n",
    "        'fragmentationCategory',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'species',\n",
    "        'numCoveredAAsDB',\n",
    "        'numNcuts'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "B721_Whole['Peptide:UID'] = B721_Whole.apply(numberPeptides, axis=1)\n",
    "B721_Whole['sequenceList'] = B721_Whole.apply(mapSequenceUpdate, axis=1)\n",
    "B721_Whole.to_csv(\n",
    "    '../../data/tsv/B721.Whole.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CLL MHC-I"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process CLL 5283"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLL_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/CLL_5283.MHCI.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "CLL_MHC = removeBlacklist(CLL_MHC)\n",
    "CLL_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'subgroupSpecific',\n",
    "        'numLCrunsObserved',\n",
    "        'missedCleavages',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'rank2Sequence',\n",
    "        'sequenceMap',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'parent_m_over_z',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'next_aa',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'type_MutationEvent',\n",
    "        'entry_name',\n",
    "        'recallMstag',\n",
    "        'numCoveredAAsDB',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'faimsCV',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'numNcuts'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "CLL_MHC['Peptide:UID'] = CLL_MHC.apply(numberPeptides, axis=1)\n",
    "CLL_MHC['sequenceList'] = CLL_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "CLL_MHC.to_csv(\n",
    "    '../../data/tsv/CLL.5283.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process CLL 5328"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLL_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/CLL_5328.MHCI.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "CLL_MHC = removeBlacklist(CLL_MHC)\n",
    "CLL_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'subgroupSpecific',\n",
    "        'numLCrunsObserved',\n",
    "        'missedCleavages',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'rank2Sequence',\n",
    "        'sequenceMap',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'parent_m_over_z',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'next_aa',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'recallMstag',\n",
    "        'numCoveredAAsDB',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'faimsCV',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'numNcuts'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "CLL_MHC['Peptide:UID'] = CLL_MHC.apply(numberPeptides, axis=1)\n",
    "CLL_MHC['sequenceList'] = CLL_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "CLL_MHC.to_csv(\n",
    "    '../../data/tsv/CLL.5328.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process CLL 5341"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLL_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/CLL_5341.MHCI.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "CLL_MHC = removeBlacklist(CLL_MHC)\n",
    "CLL_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'subgroupSpecific',\n",
    "        'numLCrunsObserved',\n",
    "        'missedCleavages',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'rank2Sequence',\n",
    "        'sequenceMap',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'parent_m_over_z',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'next_aa',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'recallMstag',\n",
    "        'numCoveredAAsDB',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'faimsCV',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'numNcuts'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "CLL_MHC['Peptide:UID'] = CLL_MHC.apply(numberPeptides, axis=1)\n",
    "CLL_MHC['sequenceList'] = CLL_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "CLL_MHC.to_csv(\n",
    "    '../../data/tsv/CLL.5341.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GBM MHC-I & Whole"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process GBM H4512 Whole"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "GBM_Whole = pandas.read_csv(\n",
    "    '../../data/ssv/GBM_H4512.Whole.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "# GBM_Whole['IFN'] = GBM_Whole.apply(wholeGBMIFN, axis=1)\n",
    "GBM_Whole.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'subgroupSpecific',\n",
    "        'numLCrunsObserved',\n",
    "        'missedCleavages',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'nterm',\n",
    "        'rank2Sequence',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'parent_m_over_z',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'sequenceMap',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'accession_number',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'next_aa',\n",
    "        'fragmentationCategory',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'recallMstag',\n",
    "        'numCoveredAAsDB',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numNCcuts',\n",
    "        'numIcuts'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "GBM_Whole['Peptide:UID'] = GBM_Whole.apply(numberPeptides, axis=1)\n",
    "GBM_Whole['sequenceList'] = GBM_Whole.apply(mapSequenceUpdate, axis=1)\n",
    "GBM_Whole.to_csv(\n",
    "    '../../data/tsv/GBM.H4512.Whole.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process GBM H4512 MHC-I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "GBM_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/GBM_H4512_BT145_20180622.peptideExport.CS.4.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "GBM_MHC = removeBlacklist(GBM_MHC)\n",
    "GBM_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'numLCrunsObserved',\n",
    "        'missedCleavages',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'rank2Sequence',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'parent_m_over_z',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'sequenceMap',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'next_aa',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'trapFillMsec',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'recallMstag',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'faimsCV',\n",
    "        'numCoveredAAsDB',\n",
    "        'refinedType',\n",
    "        'mergeType',\n",
    "        'plotType',\n",
    "        'condType'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "GBM_MHC['Peptide:UID'] = GBM_MHC.apply(numberPeptides, axis=1)\n",
    "GBM_MHC['sequenceList'] = GBM_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "GBM_MHC.to_csv(\n",
    "    '../../data/tsv/GBM.H4512.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process GBM H4512 IFN MHC-I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "GBM_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/GBM_H4512_BT145_20180622_IFN.peptideExport.CS.1.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "GBM_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'numLCrunsObserved',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'parent_m_over_z',\n",
    "        'missedCleavages',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'rank2Sequence',\n",
    "        'numCoveredAAsDB',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'sequenceMap',\n",
    "        'refinedType',\n",
    "        'mergeType',\n",
    "        'plotType',\n",
    "        'condType'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "GBM_MHC['Peptide:UID'] = GBM_MHC.apply(numberPeptides, axis=1)\n",
    "GBM_MHC['sequenceList'] = GBM_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "GBM_MHC.to_csv(\n",
    "    '../../data/tsv/GBM.H4512.IFN.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process GBM H4198 MHC-I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "GBM_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/GBM_H4198_BT187_20180622.peptideExport.CS.3.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "GBM_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'numLCrunsObserved',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'parent_m_over_z',\n",
    "        'missedCleavages',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'rank2Sequence',\n",
    "        'numCoveredAAsDB',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'sequenceMap',\n",
    "        'refinedType',\n",
    "        'mergeType',\n",
    "        'plotType',\n",
    "        'condType'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "GBM_MHC['Peptide:UID'] = GBM_MHC.apply(numberPeptides, axis=1)\n",
    "GBM_MHC['sequenceList'] = GBM_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "GBM_MHC.to_csv(\n",
    "    '../../data/tsv/GBM.H4198.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process GBM H4198 IFN MHC-I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "GBM_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/GBM_H4198_BT187_20180622_IFN.peptideExport.CS.1.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "GBM_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'numLCrunsObserved',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'parent_m_over_z',\n",
    "        'missedCleavages',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'rank2Sequence',\n",
    "        'numCoveredAAsDB',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'sequenceMap',\n",
    "        'refinedType',\n",
    "        'mergeType',\n",
    "        'plotType',\n",
    "        'condType'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "GBM_MHC['Peptide:UID'] = GBM_MHC.apply(numberPeptides, axis=1)\n",
    "GBM_MHC['sequenceList'] = GBM_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "GBM_MHC.to_csv(\n",
    "    '../../data/tsv/GBM.H4198.IFN.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process GBM 7 MHC-I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "GBM_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/GBM07_14362-007_20170912.peptideExport.CS.2.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "GBM_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'numLCrunsObserved',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'parent_m_over_z',\n",
    "        'missedCleavages',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'rank2Sequence',\n",
    "        'numCoveredAAsDB',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'sequenceMap',\n",
    "        'refinedType',\n",
    "        'mergeType',\n",
    "        'plotType',\n",
    "        'condType',\n",
    "        'type_MutationEvent'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "GBM_MHC['Peptide:UID'] = GBM_MHC.apply(numberPeptides, axis=1)\n",
    "GBM_MHC['sequenceList'] = GBM_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "GBM_MHC.to_csv(\n",
    "    '../../data/tsv/GBM.7.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process GBM 7 IFN MHC-I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "GBM_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/GBM07_14362-007_20170912_withIFN.peptideExport.CS.1.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "GBM_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'numLCrunsObserved',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'parent_m_over_z',\n",
    "        'missedCleavages',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'rank2Sequence',\n",
    "        'numCoveredAAsDB',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'sequenceMap',\n",
    "        'refinedType',\n",
    "        'mergeType',\n",
    "        'plotType',\n",
    "        'condType',\n",
    "        'type_MutationEvent'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "GBM_MHC['Peptide:UID'] = GBM_MHC.apply(numberPeptides, axis=1)\n",
    "GBM_MHC['sequenceList'] = GBM_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "GBM_MHC.to_csv(\n",
    "    '../../data/tsv/GBM.7.IFN.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Melanoma MHC-I"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process Mel 2 MHC-I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "MEL_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/Mel2.MHCI.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "MEL_MHC = MEL_MHC[MEL_MHC['directory'] != 'Mel02_13240-002_20180803_4IPs']\n",
    "MEL_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'numLCrunsObserved',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'parent_m_over_z',\n",
    "        'missedCleavages',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'rank2Sequence',\n",
    "        'numCoveredAAsDB',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'sequenceMap',\n",
    "        'type_MutationEvent',\n",
    "        'subgroupSpecific'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "MEL_MHC['Peptide:UID'] = MEL_MHC.apply(numberPeptides, axis=1)\n",
    "MEL_MHC['sequenceList'] = MEL_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "MEL_MHC.to_csv(\n",
    "    '../../data/tsv/MEL.2.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process Mel 2 10IP MHC-I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "MEL_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/Mel02_13240-002_20190419_10IPs.peptideExport.CS.1.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "MEL_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'numLCrunsObserved',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'parent_m_over_z',\n",
    "        'missedCleavages',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'rank2Sequence',\n",
    "        'numCoveredAAsDB',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'sequenceMap',\n",
    "        'type_MutationEvent',\n",
    "        'refinedType',\n",
    "        'mergeType',\n",
    "        'plotType',\n",
    "        'condType'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "MEL_MHC['Peptide:UID'] = MEL_MHC.apply(numberPeptides, axis=1)\n",
    "MEL_MHC['sequenceList'] = MEL_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "MEL_MHC.to_csv(\n",
    "    '../../data/tsv/MEL.2.10IP.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process Mel 2 SEL MHC-I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "MEL_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/Mel02_13240-002_20190108_MEL_sel.peptideExport.CS.1.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "MEL_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'numLCrunsObserved',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'parent_m_over_z',\n",
    "        'missedCleavages',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'rank2Sequence',\n",
    "        'numCoveredAAsDB',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'sequenceMap',\n",
    "        'type_MutationEvent',\n",
    "        'refinedType',\n",
    "        'mergeType',\n",
    "        'plotType',\n",
    "        'condType'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "MEL_MHC['Peptide:UID'] = MEL_MHC.apply(numberPeptides, axis=1)\n",
    "MEL_MHC['sequenceList'] = MEL_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "MEL_MHC.to_csv(\n",
    "    '../../data/tsv/MEL.2s.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process Mel 2 SEL IFN MHC-I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "MEL_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/Mel02_13240-002_20190108_MEL_sel_IFN.peptideExport.CS.1.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "MEL_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'numLCrunsObserved',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'parent_m_over_z',\n",
    "        'missedCleavages',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'rank2Sequence',\n",
    "        'numCoveredAAsDB',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'sequenceMap',\n",
    "        'type_MutationEvent',\n",
    "        'refinedType',\n",
    "        'mergeType',\n",
    "        'plotType',\n",
    "        'condType'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "MEL_MHC['Peptide:UID'] = MEL_MHC.apply(numberPeptides, axis=1)\n",
    "MEL_MHC['sequenceList'] = MEL_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "MEL_MHC.to_csv(\n",
    "    '../../data/tsv/MEL.2s.IFN.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process Mel 6 MHC-I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "MEL_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/Mel06_13240-006_20180622.peptideExport.CS.2.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "MEL_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'numLCrunsObserved',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'parent_m_over_z',\n",
    "        'missedCleavages',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'rank2Sequence',\n",
    "        'numCoveredAAsDB',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'sequenceMap',\n",
    "        'refinedType',\n",
    "        'mergeType',\n",
    "        'plotType',\n",
    "        'condType'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "MEL_MHC['Peptide:UID'] = MEL_MHC.apply(numberPeptides, axis=1)\n",
    "MEL_MHC['sequenceList'] = MEL_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "MEL_MHC.to_csv(\n",
    "    '../../data/tsv/MEL.6.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process Mel 6 IFN MHC-I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "MEL_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/Mel06_13240-006_20180622_IFN.peptideExport.CS.1.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "MEL_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'numLCrunsObserved',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'parent_m_over_z',\n",
    "        'missedCleavages',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'rank2Sequence',\n",
    "        'numCoveredAAsDB',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'sequenceMap',\n",
    "        'refinedType',\n",
    "        'mergeType',\n",
    "        'plotType',\n",
    "        'condType'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "MEL_MHC['Peptide:UID'] = MEL_MHC.apply(numberPeptides, axis=1)\n",
    "MEL_MHC['sequenceList'] = MEL_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "MEL_MHC.to_csv(\n",
    "    '../../data/tsv/MEL.6.IFN.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process Mel 11 MHC-I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "MEL_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/Mel11_13240-011_2018_2016.peptideExport.CS.4.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "MEL_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'numLCrunsObserved',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'parent_m_over_z',\n",
    "        'missedCleavages',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'rank2Sequence',\n",
    "        'numCoveredAAsDB',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'sequenceMap',\n",
    "        'type_MutationEvent',\n",
    "        'refinedType',\n",
    "        'mergeType',\n",
    "        'plotType',\n",
    "        'condType'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "MEL_MHC['Peptide:UID'] = MEL_MHC.apply(numberPeptides, axis=1)\n",
    "MEL_MHC['sequenceList'] = MEL_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "MEL_MHC.to_csv(\n",
    "    '../../data/tsv/MEL.11.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process Mel 11 IFN MHC-I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "MEL_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/Mel11_13240-011_20180308_IFN.peptideExport.CS.1.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "MEL_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'numLCrunsObserved',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'parent_m_over_z',\n",
    "        'missedCleavages',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'rank2Sequence',\n",
    "        'numCoveredAAsDB',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'sequenceMap',\n",
    "        'type_MutationEvent',\n",
    "        'refinedType',\n",
    "        'mergeType',\n",
    "        'plotType',\n",
    "        'condType'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "MEL_MHC['Peptide:UID'] = MEL_MHC.apply(numberPeptides, axis=1)\n",
    "MEL_MHC['sequenceList'] = MEL_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "MEL_MHC.to_csv(\n",
    "    '../../data/tsv/MEL.11.IFN.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process Mel 15 MHC-I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "MEL_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/Mel15_13240-015_2018_2016.peptideExport.CS.2.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "MEL_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'numLCrunsObserved',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'parent_m_over_z',\n",
    "        'missedCleavages',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'rank2Sequence',\n",
    "        'numCoveredAAsDB',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'sequenceMap',\n",
    "        'refinedType',\n",
    "        'mergeType',\n",
    "        'plotType',\n",
    "        'condType'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "MEL_MHC['Peptide:UID'] = MEL_MHC.apply(numberPeptides, axis=1)\n",
    "MEL_MHC['sequenceList'] = MEL_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "MEL_MHC.to_csv(\n",
    "    '../../data/tsv/MEL.15.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process Mel 15 IFN MHC-I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "MEL_MHC = pandas.read_csv(\n",
    "    '../../data/ssv/Mel15_13240-015_2018_IFN.peptideExport.CS.1.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "MEL_MHC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'numLCrunsObserved',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'parent_m_over_z',\n",
    "        'missedCleavages',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'rank2Sequence',\n",
    "        'numCoveredAAsDB',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'sequenceMap',\n",
    "        'refinedType',\n",
    "        'mergeType',\n",
    "        'plotType',\n",
    "        'condType'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "MEL_MHC['Peptide:UID'] = MEL_MHC.apply(numberPeptides, axis=1)\n",
    "MEL_MHC['sequenceList'] = MEL_MHC.apply(mapSequenceUpdate, axis=1)\n",
    "MEL_MHC.to_csv(\n",
    "    '../../data/tsv/MEL.15.IFN.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ovarian Cancer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### OV CP-594"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "OV = pandas.read_csv(\n",
    "    '../../data/ssv/OV.CP-594.MHCI.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "OV.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'subgroupSpecific',\n",
    "        'numLCrunsObserved',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'parent_m_over_z',\n",
    "        'missedCleavages',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'rank2Sequence',\n",
    "        'numCoveredAAsDB',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'sequenceMap'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "OV['Peptide:UID'] = OV.apply(numberPeptides, axis=1)\n",
    "OV['sequenceList'] = OV.apply(mapSequenceUpdate, axis=1)\n",
    "OV.to_csv(\n",
    "    '../../data/tsv/OV.CP-594.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### OV SLS3-M1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "OV = pandas.read_csv(\n",
    "    '../../data/ssv/OV.SLS3-M1.MHCI.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "OV.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'subgroupSpecific',\n",
    "        'numLCrunsObserved',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'parent_m_over_z',\n",
    "        'missedCleavages',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'rank2Sequence',\n",
    "        'numCoveredAAsDB',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'sequenceMap'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "OV['Peptide:UID'] = OV.apply(numberPeptides, axis=1)\n",
    "OV['sequenceList'] = OV.apply(mapSequenceUpdate, axis=1)\n",
    "OV.to_csv(\n",
    "    '../../data/tsv/OV.SLS3-M1.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Renal Cell Carcinoma"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### RCC 9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "RCC = pandas.read_csv(\n",
    "    '../../data/ssv/RCC.9.MHCI.ssv',\n",
    "    sep=';',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "RCC.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'subgroupSpecific',\n",
    "        'numLCrunsObserved',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'accession_numbers',\n",
    "        'geneSymbol',\n",
    "        'orfCategory',\n",
    "        'Ensembl_GENCODE_GeneID_v19',\n",
    "        'geneSymbol_RibORF',\n",
    "        'type_RibORF',\n",
    "        'entry_name',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'accession_number',\n",
    "        'StartAA',\n",
    "        'previous_aa',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'trapFillMsec',\n",
    "        'faimsCV',\n",
    "        'parent_m_over_z',\n",
    "        'missedCleavages',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'fragmentationCategory',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'rank2Sequence',\n",
    "        'numCoveredAAsDB',\n",
    "        'num_seqs_past_parent_filter',\n",
    "        'sequenceMap'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "RCC['Peptide:UID'] = RCC.apply(numberPeptides, axis=1)\n",
    "RCC['sequenceList'] = RCC.apply(mapSequenceUpdate, axis=1)\n",
    "RCC.to_csv(\n",
    "    '../../data/tsv/RCC.9.MHCI.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Database Comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "DBC_RNA = pandas.read_csv(\n",
    "    '../../data/csv/DBC.RNA.csv',\n",
    "    sep=',',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "DBC_RNA = removeBlacklist(DBC_RNA)\n",
    "DBC_RNA['allele'] = DBC_RNA['directory'].apply(B721Allele)\n",
    "DBC_RNA.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'subgroupSpecific',\n",
    "        'numLCrunsObserved',\n",
    "        'missedCleavages',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'StartAA',\n",
    "        'rank2Sequence',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'parent_m_over_z',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'sequenceMap',\n",
    "        'previous_aa',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'fragmentationCategory',\n",
    "        'numCoveredAAsDB',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'entry_name'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "DBC_RNA['Peptide:UID'] = DBC_RNA.apply(numberPeptides, axis=1)\n",
    "DBC_RNA['sequenceList'] = DBC_RNA.apply(mapSequenceUpdate, axis=1)\n",
    "DBC_RNA.to_csv(\n",
    "    '../../data/tsv/DBC.RNA.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process RPF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "DBC_RPF = pandas.read_csv(\n",
    "    '../../data/csv/DBC.RPF.csv',\n",
    "    sep=',',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "DBC_RPF = removeBlacklist(DBC_RPF)\n",
    "DBC_RPF['allele'] = DBC_RPF['directory'].apply(B721Allele)\n",
    "DBC_RPF.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'subgroupSpecific',\n",
    "        'numLCrunsObserved',\n",
    "        'missedCleavages',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'StartAA',\n",
    "        'rank2Sequence',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'parent_m_over_z',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'sequenceMap',\n",
    "        'previous_aa',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'fragmentationCategory',\n",
    "        'numCoveredAAsDB',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'entry_name'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "DBC_RPF['Peptide:UID'] = DBC_RPF.apply(numberPeptides, axis=1)\n",
    "DBC_RPF['sequenceList'] = DBC_RPF.apply(mapSequenceUpdate, axis=1)\n",
    "DBC_RPF.to_csv(\n",
    "    '../../data/tsv/DBC.RPF.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process B721"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "DBC_B721 = pandas.read_csv(\n",
    "    '../../data/csv/DBC.B721.csv',\n",
    "    sep=',',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "DBC_B721 = removeBlacklist(DBC_B721)\n",
    "DBC_B721['allele'] = DBC_B721['directory'].apply(B721Allele)\n",
    "DBC_B721.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'subgroupSpecific',\n",
    "        'numLCrunsObserved',\n",
    "        'missedCleavages',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'StartAA',\n",
    "        'rank2Sequence',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'parent_m_over_z',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'sequenceMap',\n",
    "        'previous_aa',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'fragmentationCategory',\n",
    "        'numCoveredAAsDB',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'entry_name'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "DBC_B721['Peptide:UID'] = DBC_B721.apply(numberPeptides, axis=1)\n",
    "DBC_B721['sequenceList'] = DBC_B721.apply(mapSequenceUpdate, axis=1)\n",
    "DBC_B721.to_csv(\n",
    "    '../../data/tsv/DBC.B721.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process Pan Sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "DBC_PS = pandas.read_csv(\n",
    "    '../../data/csv/DBC.PS.csv',\n",
    "    sep=',',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "DBC_PS = removeBlacklist(DBC_PS)\n",
    "DBC_PS['allele'] = DBC_PS['directory'].apply(B721Allele)\n",
    "DBC_PS.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'subgroupSpecific',\n",
    "        'numLCrunsObserved',\n",
    "        'missedCleavages',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'StartAA',\n",
    "        'rank2Sequence',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'parent_m_over_z',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'sequenceMap',\n",
    "        'previous_aa',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'fragmentationCategory',\n",
    "        'numCoveredAAsDB',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'entry_name'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "DBC_PS['Peptide:UID'] = DBC_PS.apply(numberPeptides, axis=1)\n",
    "DBC_PS['sequenceList'] = DBC_PS.apply(mapSequenceUpdate, axis=1)\n",
    "DBC_PS.to_csv(\n",
    "    '../../data/tsv/DBC.PS.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process Null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "DBC_NULL = pandas.read_csv(\n",
    "    '../../data/csv/DBC.NULL.csv',\n",
    "    sep=',',\n",
    "    header=0,\n",
    "    index_col=False)\n",
    "DBC_NULL = removeBlacklist(DBC_NULL)\n",
    "DBC_NULL['allele'] = DBC_NULL['directory'].apply(B721Allele)\n",
    "DBC_NULL.drop(\n",
    "    [\n",
    "        'number',\n",
    "        'directory',\n",
    "        'parent_charge',\n",
    "        'subgroupSpecific',\n",
    "        'numLCrunsObserved',\n",
    "        'missedCleavages',\n",
    "        'numKorR',\n",
    "        'totalIntensity',\n",
    "        'variableSites',\n",
    "        'StartAA',\n",
    "        'rank2Sequence',\n",
    "        'cys',\n",
    "        'modifications',\n",
    "        'precursorAveragineChiSquared',\n",
    "        'precursorIsolationPurityPercent',\n",
    "        'chromatographicPeakWidthSec',\n",
    "        'parent_m_over_z',\n",
    "        'matched_parent_mass',\n",
    "        'delta_parent_mass',\n",
    "        'delta_parent_mass_ppm',\n",
    "        'peptide_pI',\n",
    "        'deltaRank1Rank2Score',\n",
    "        'sequenceMap',\n",
    "        'previous_aa',\n",
    "        'next_aa',\n",
    "        'recallMstag',\n",
    "        'fragmentationCategory',\n",
    "        'numCoveredAAsDB',\n",
    "        'numNcuts',\n",
    "        'numCcuts',\n",
    "        'numIcuts',\n",
    "        'numNCcuts',\n",
    "        'protein_mw',\n",
    "        'species',\n",
    "        'entry_name'\n",
    "    ],\n",
    "    axis=1,\n",
    "    inplace=True)\n",
    "DBC_NULL['Peptide:UID'] = DBC_NULL.apply(numberPeptides, axis=1)\n",
    "DBC_NULL['sequenceList'] = DBC_NULL.apply(mapSequenceUpdate, axis=1)\n",
    "DBC_NULL.to_csv(\n",
    "    '../../data/tsv/DBC.NULL.tsv',\n",
    "    sep='\\t',\n",
    "    header=True,\n",
    "    index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
